src/election.ts

```diff
@@ -0,0 +1,282 @@
+  }
+
+  public getPrefix() {
+    return `${Election.prefix}/${this.name}/`;
+  }
+
+  private async waitForElected() {
+    // find last create before this
+    const lastRevision = new BigNumber(this.leaderRevision).minus(1).toString();
+    const result = await this.namespace
+      .getAll()
+      .maxCreateRevision(lastRevision)
+      .sort('Create', 'Descend')
+      .keys();
+
+    // no one before this, elected
+    if (result.length === 0) {
+      return;
+    }
+
+    // wait all keys created ealier are deleted
+    await waitForDeletes(this.namespace, result);
+  }
+
+  private async observe() {
+    if (this._isObserving) {
+      return;
+    }
+
+    try {
+      this._isObserving = true;
+
+      // looking for current leader
+      let leaderKey = '';
+      const result = await this.namespace.getAll().sort('Create', 'Ascend').keys();
+
+      if (result.length === 0) {
+        // if not found, wait for leader
+        const watcher = await this.parent.watch().prefix(this.getPrefix()).create();
+        try {
+          leaderKey = await new Promise<string>((resolve, reject) => {
+            watcher.on('put', kv => resolve(kv.key.toString()));
+            watcher.on('error', reject);
+          });
+        } finally {
+          await watcher.cancel();
+        }
+      } else {
+        leaderKey = `${this.getPrefix()}${result[0]}`;
+      }
+
+      // emit current leader
+      this.emit('leader', leaderKey);
+
+      // wait for delete event
+      await waitForDelete(this.parent, leaderKey);
+    } finally {
+      this._isObserving = false;
+    }
+
+    // only keep watch if listened
+    if (this.listenerCount('leader') > 0) {
+      this.tryObserve();
+    }
+  }
+
+  private tryObserve(): void {
+    this.observe().catch(error => this.emitOrThrowError(error));
+  }
+
+  private shouldObserve(event: string|symbol): boolean {
+    switch (event) {
+      case 'leader': return true;
+      default: return false;
+    }
+  }
+
+  private emitOrThrowError(error: any): void {
+    if (this.listenerCount('error') > 0) {
+      this.emit('error', error);
+    } else {
+      throw error;
+    }
+  }
+
+  private onLeaseLost() {
+    this.lease = this.namespace.lease(this.ttl);
+    this.lease.on('lost', () => this.onLeaseLost());
+    this.leaseId = '';
+  }
+}
+
+async function waitForDelete(namespace: Namespace, key: string) {
+  const watcher = await namespace.watch().key(key).create();
+  const deleteOrError = new Promise(async (resolve, reject) => {
+    // waiting for deleting of that key
+    watcher.on('delete', resolve);
+    watcher.on('error', reject);
+  });
+
+  try {
+    await deleteOrError;
+  } finally {
+    await watcher.cancel();
+  }
+}
+
+async function waitForDeletes(namespace: Namespace, keys: string[]) {
+  if (keys.length === 0) {
+    return;
+  } else if (keys.length === 1) {
+    return waitForDelete(namespace, keys[0]);
+  }
+
+  const tasks = keys.map(key => async () => {
```

Comment:
Do these need to be strictly executed in order? Otherwise you could use a `for of` loop instead.